pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -0,0 +1,212 @@
+) *PruneController {
+	c := &PruneController{
+		targetNamespace:        targetNamespace,
+		staticPodName:          staticPodName,
+		command:                command,
+		failedRevisionLimit:    failedRevisionLimit,
+		succeededRevisionLimit: succeededRevisionLimit,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventRecorder,
+
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PruneController"),
+		prunerPodImageFn: getPrunerPodImageFromEnv,
+	}
+
+	return c
+}
+
+func (c *PruneController) pruneRevisionHistory(resourceDir, configMapDir string) error {
+	var succeededRevisionIDs, failedRevisionIDs []string
+
+	// Walk the resource directory looking for files that match our static pod name prefix
+	err := filepath.Walk(resourceDir, func(currentPath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() && strings.HasPrefix(info.Name(), c.staticPodName) {
+			// Read the phase configmap for this revision and sort into failed or succeeded
+			phase, err := ioutil.ReadFile(path.Join(resourceDir, configMapDir))
+			if err != nil {
+				return err
+			}
+			switch string(phase) {
+			case string(corev1.PodSucceeded):
+				succeededRevisionIDs = append(succeededRevisionIDs, info.Name())
+			case string(corev1.PodFailed):
+				failedRevisionIDs = append(failedRevisionIDs, info.Name())
+			}
+		} else {
+			return filepath.SkipDir
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Make sure the lists are sorted in a way that accounts for numbering
+	sort.Sort(sortorder.Natural(succeededRevisionIDs))
+	sort.Sort(sortorder.Natural(failedRevisionIDs))
+
+	// Get list of protected IDs
+	protectedSucceededRevisionIDs := succeededRevisionIDs[len(succeededRevisionIDs)-c.succeededRevisionLimit:]
+	protectedFailedRevisionIDs := failedRevisionIDs[len(failedRevisionIDs)-c.failedRevisionLimit:]
+	excludedIDs := append(protectedSucceededRevisionIDs, protectedFailedRevisionIDs...)
+	sort.Sort(sortorder.Natural(excludedIDs))
+
+	return c.ensurePrunePod(prunerPodName, excludedIDs[len(excludedIDs)-1], excludedIDs)
+}
+
+func (c *PruneController) ensurePrunePod(podName, maxEligibleRevision string, protectedRevisions []string) error {
+	required := resourceread.ReadPodV1OrDie([]byte(prunerPod))
+	required.Name = podName
+	required.Namespace = c.targetNamespace
+	required.Spec.Containers[0].Image = c.prunerPodImageFn()
+	required.Spec.Containers[0].Command = c.command
+	required.Spec.Containers[0].Args = append(required.Spec.Containers[0].Args,
+		fmt.Sprintf("-v=%d", 4),
+		fmt.Sprintf("--max-eligible-id=%s", maxEligibleRevision),
+		fmt.Sprintf("--protected-ids=%s", strings.Join(protectedRevisions, ",")),
+		fmt.Sprintf("--resource-dir=%s", defaultResourceDir),
+		fmt.Sprintf("--static-pod-name=%s", c.staticPodName),
+	)
+
+	if _, err := c.kubeClient.CoreV1().Pods(c.targetNamespace).Create(required); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
I expected you to need to pin this to the "correct" node and run three of them.

Reply:
> I expected you to need to pin this to the "correct" node and run three of them.

and run one for every node

Reply:
added here: https://github.com/openshift/library-go/pull/82/files#diff-5c9aefda4344c1bce9b242c5c5f5d82eR114